nixos/modules/services/x11/desktop-managers/xfce4-14.nix

```diff
@@ -0,0 +1,156 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.xserver.desktopManager.xfce4-14;
+in
+
+{
+  # added 2019-08-18
+  # needed to preserve some semblance of UI familarity
+  # with original XFCE module
+  imports = [
+    (mkRenamedOptionModule
+      [ "services" "xserver" "desktopManager" "xfce4-14" "extraSessionCommands" ]
+      [ "services" "xserver" "displayManager" "sessionCommands" ])
+  ];
+
+  options = {
+    services.xserver.desktopManager.xfce4-14 = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Enable the Xfce desktop environment.";
+      };
+
+    # TODO: support thunar plugins
+    #   thunarPlugins = mkOption {
+    #     default = [];
+    #     type = types.listOf types.package;
+    #     example = literalExample "[ pkgs.xfce4-14.thunar-archive-plugin ]";
+    #     description = ''
+    #       A list of plugin that should be installed with Thunar.
+    #     '';
+    #  };
+
+      noDesktop = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Don't install XFCE desktop components (xfdesktop, panel and notification daemon).";
+      };
+
+      enableXfwm = mkOption {
+        type = types.bool;
+        default = true;
+        description = "Enable the XFWM (default) window manager.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      { assertion = (config.services.xserver.desktopManager.xfce.enable == false);
+        message = "You cannot enable both Xfce 4.12 and Xfce 4.14. Please use either services.xserver.desktopManager.xfce4-14.enable or services.xserver.desktopManager.xfce.enable standalone.";
```

Comment:
They may work being both enabled, `lightdm` or `slim` will ask which one to use

Reply:
The concern is mostly trying to start session where the last was 4.14 and then logging in as 4.12.
I'm assuming that some sort of migration happens for 4.12 -> 4.14 but no reversion.

Reply:
The `assertion` does not prevent downgrade, only makes it less convinient (via `nixos-rebuild`).

Reply:
Oh it prevents both being enabled but doesn't succinctly prevent the downgrade from occurring.
As someone could not have them both enabled but still switch between them with successive rebuilds. I guess that's something we'd have to allow as I don't think there's a way to track that?

Reply:
May it be possible to add a tag "version" instead ?
looks like:
```
services.xserver.desktopManager.xfce = {
  enable = true;
  version = 4.14;
};
```
I'm feeling this to be clearer and if possible, a kind of "standard" able for other kind of config stuff to choose version with.

Reply:
4.12 is already removed, it was an exceptional period of having two versions XFCE in nixpkgs

Reply:
@conferno yes sure, i'm pretty new nixos user, but i do understand it.
But this situation, i think, is a very good experience to show that this can happen (and will happen again).
I also don't understand if it is possible to increase the code to get this requirement to add version selection on this way (who seems to be clearer).
The subject i'm trying to talking about is not to define if it is an exception but if it is possible to to add a clear method to select a version of a package and much more, a group of packages. And by the way to add a "version" tag control inside.

1. Possible or not ? (and why)

1. Good idea or not (and why) ?

The idea is born, because the situation (exceptional or not) exist.

Reply:
It would be possible but it would be extra maintenance burden. We generally only want to support a single version at the time.

The reason we introduced 4.13 (development snapshot of 4.14) was because 4.12 was using legacy technologies but 4.14 has not been released yet. Hopefully, this situation will not repeat very often.

Reply:
> And by the way to add a "version" tag control inside.
> 
> 1. Possible or not ? (and why)
> 2. Good idea or not (and why) ?
> 
> The idea is born, because the situation (exceptional or not) exist.

I had both 4.12 and 4.14 installed, choosing between them on the login screen.
It is the same as having installed both Gnome and Plasma.
Adding version tag makes it impossible to install more than 1 version

Reply:
@volth it should be possible to increase multi versions to be installed by this concept:
```
version = [ 4.12 4.14 ];
```
@jtojnar i understand better your explication on that point. It is then a choice (maybe a good choice to for most of the situations/users). I'm thinking that what can do the maximum can also choose to do the minimum, but if the limitation is minimum, we can not do the maximum.
It is an idea, but i'm not NixOS dev, so i can not push more ahead this idea (i also don't know devs implication to go to maintain this kind of quick idea who should be not adapted to the situation).
I understand better, thank you for your explication around this subject.
I can see that if i do en error like:
```
nix-env -S 0
```
i break the user config. And the solution is to delete ~/.nix-profile symbolic link and re-start.
So i can see that there is maybe many other priorities to fix or to go for TDD codes for make NixOS stronger out of the box. 
I still follow the evolution of NixOS because i see this distro to have many very good ideas. 
Thank you for your great job.